packages/hash/api/src/lib/schemas/jsonSchema.ts

```diff
@@ -138,9 +225,12 @@ export class JsonSchemaCompiler {
    * @param schema schema object
    * @throws if any types of duplicate properties mismatch
    */
-  async prevalidateProperties(schema: any) {
+  async prevalidateProperties(schema: any): Promise<JSONSchema> {
     const self = this;
-    const bundled = await $RefParser.bundle(schema, {
+    const dereferences = await $RefParser.dereference(schema, {
```

Comment:
@CiaranMn I've already discussed this with Ahmad internally, but wanted to start a thread on this where you could give your thoughts on the topic (if you have any).

I have concerns about using a JSON-Schema dereferences in this way, because of how I understand they are resolving references (`$ref: ...`) and how I understand we plan on using references in our schemas.

Just to make sure we’re on the same page, currently the two places where we are using the `$ref` keyword in our schemas are:
1. to declare the “parent” schemas using the using the `allOf` array
2. to declare an outgoing links from one schema to another

If this assumption is correct, using traditional packages designed to resolve references in JSON-Schemas (such as `$RefParser.dereference(…)`, or `ajv`) can cause issues for us, as they assume they should recursively resolve all (acyclic) references they encounter in a schema. This assumption is made because traditional JSON-Schemas are used to define the structure of a _finite_ JSON object, where references can be used to separate complex definitions of that JSON object.

However, we are not using JSON-Schema references to define the structure of a single JSON object - we’re using the references to define an ontology between schemas. This means that our collection of JSON-Schemas could be considered as a large circular directed graph, which should never be fully resolved at once (an assumption `$RefParser.dereference(…)` and all other JSON-Schema libraries I’ve come across make).

So if for instance we were to use `$RefParser.dereference` to get all the parents of an entity type (as we are currently doing in this PR), not only would this operation fetch and resolve all the references made in the `allOf` array, it would also resolve all the outgoing linked schemas, and their outgoing linked schemas, and their parent schemas, and their grand parent schemas, and so on. This won't scale.

What we would need from an existing JSON-Schema dereferencer is the ability to tell it “please don’t bother resolving references you find here, and only resolve the references made in the `allOf` array” - but as far as I’m able to tell this is not an option supported by `$RefParser.dereference` (or by other JSON-Schema parsers I’ve come across).

Therefore I think we will instead have to resolve references ourselves. For instance for the functionality required in this PR, we could recursively call our existing `getEntityTypeParents` db method, which manually looks up references made in the `allOf` array on the database level.

Let me know if you have any thoughts on this @CiaranMn , and whether you agree with the observations/claims I made here.

Reply:
Just chiming in to add that the current schema-compatibility checking is agnostic to how we fetch json schemas and their properties. So moving away from relying on existing json schema libraries wouldn't require re-working that part of the code.

Reply:
Yes, I agree that we do not want the default behavior of JSON validator libraries of attempting to resolve every onward $ref in a schema.

A couple of additional points which will be important in writing our own:

There is a third way in which $refs are used in our schemas: to refer to sub-schemas stored in [$defs](https://json-schema.org/understanding-json-schema/structuring.html#defs) - we do want to resolve these sub-schemas (because they are important to generating a complete list of the properties in the schema). We should identify and handle these
A robust $ref parser would handle relative URLs as well as absolute ones - the relative URLs are resolved against the same base as the $id.

Reply:
In that case I will close the PR as it will need to look very different. I'll try to capture what was mentioned here in the asana task and take it from there

Reply:
Understood, apologies @thehabbos007 for not having caught this earlier, and for our non-traditional usage of the `$ref` field not having been made clear in the task descriptions/prior discussions.